### PR TITLE
Refactor XmlWriter{Async}

### PR DESCRIPTION
**Pull request by kronic**

### CODE REVIEW

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlWriter.cs [hunk 1]
```diff
@@ -460,7 +426,7 @@ public virtual void WriteNode(XmlReader reader, bool defattr)
                         WriteFullEndElement();
                         break;
                 }
-            } while (reader.Read() && (d < reader.Depth || (d == reader.Depth && reader.NodeType == XmlNodeType.EndElement)));
+            } while (reader.Read() && (d < reader.Depth || d == reader.Depth && reader.NodeType == XmlNodeType.EndElement));
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Core/XmlWriter.cs:429]: Please revert this.  It's not clear to most readers that && has higher precedence than ||.

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlWriter.cs [hunk 2]
```diff
@@ -490,7 +456,7 @@ public virtual void WriteNode(XPathNavigator navigator, bool defattr)
                             do
                             {
                                 IXmlSchemaInfo? schemaInfo = navigator.SchemaInfo;
-                                if (defattr || (schemaInfo == null || !schemaInfo.IsDefault))
+                                if (defattr || schemaInfo is not { IsDefault: true })
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Core/XmlWriter.cs:459]: I don't think pattern matching in a case like this is any more clear. Please revert.

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlWriterAsync.cs [hunk 3]
```diff
@@ -466,7 +447,7 @@ public virtual async Task WriteNodeAsync(XPathNavigator navigator, bool defattr)
                             do
                             {
                                 IXmlSchemaInfo? schemaInfo = navigator.SchemaInfo;
-                                if (defattr || (schemaInfo == null || !schemaInfo.IsDefault))
+                                if (defattr || schemaInfo is not { IsDefault: true })
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Core/XmlWriterAsync.cs:450]: Same as earlier

**Review by stephentoub — APPROVED** [review]
Other than my comments, looks fine to me.

**Comment by kronic** [comment]
@stephentoub done

**Review by krwq — APPROVED** [review]

**Pull request merged**